### DISCUSSION TITLE
[@types/react] Derive at least a bit of information for SyntheticEvent.target

### DISCUSSION THREAD

**Discussion by flying-sheep** [DefinitelyTyped/DefinitelyTyped#56929] [Issues with a @types package]
There was a big back and forth some years ago, because while `currentTarget` is typeable, `target` isn’t, as the documentation reads

> If you thought this should be `EventTarget & T`, see https://github.com/DefinitelyTyped/DefinitelyTyped/issues/11508#issuecomment-256045682

But I’m pretty sue we do know more about `target` that just it being a `EventTarget`: If it’s a child of a HTMLElement, it’s a HTMLElement itself, right?

Based on the suggestion in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/12239#issuecomment-408684894, we could do something like:

```ts
interface SyntheticEvent<T = Element> {
  /* ... */
  target: EventTarget & (
            T extends HTMLElement
            ? (T extends EmptyHTMLElement ? T : HTMLElement)
            : {}
          )
}
```

[a comment by an automated account is not shown]

**Comment by smichelForHire** [comment]
The lack of this type means the current more-specific types like `React.ChangeEvent`, which *do* narrow their `.target`, are incompatible with others, like `React.FormEvent`, that do not narrow it. For example (**[playground link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nAwgCwIYB2A5igKIBuKBMANHAGLQgVXwC+cAZlBCHAORQUeAMYx+AbgBQUzgFcCY4BAJcIEABQpK1AFwMmLanXwEAJgBsUUfVv25CJIzACUcALwA+OOQjAzbuhScHCmltZaOq7SbFJAA)**):
```ts
import type { ChangeEvent, FormEvent } from 'react';

function foo(event: FormEvent, handler: (e: ChangeEvent) => void) {
  handler(event);
  //      ^ Type 'EventTarget' is missing the following properties from type 'Element': attributes, classList, className, clientHeight, and 161 more.
}

```

In the meantime, a workaround is to add this to a .d.ts file.  I don't love patching the core React types, but if you need it…
```ts
declare namespace React {
  interface SyntheticEvent<T = Element, E = Event> extends BaseSyntheticEvent<E, EventTarget & T, EventTarget> {
    target: EventTarget & Element;
  }
}
```

For some reason this did not work (note the final type param):
```ts
declare namespace React {
  interface SyntheticEvent<T = Element, E = Event> extends BaseSyntheticEvent<E, EventTarget & T, EventTarget & Element> {}
}
```
